www/js/lib/zimArchive.js

```diff
@@ -142,31 +142,74 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
+        // NB duplicates are removed before processing search array
+        var startArray = [];
+        // Ensure a search is done on the string exactly as typed
+        startArray.push(localPrefix);
+        // Normalize any spacing and make string all lowercase
+        var prefix = localPrefix.replace(/\s+/g, ' ').toLocaleLowerCase();
+        // Add lowercase string with initial uppercase (this is a very common pattern)
+        startArray.push(prefix.replace(/^./, function (m) {
+            return m.toLocaleUpperCase();
+        }));
+        // Get the full array of combinations to check number of combinations
+        var fullCombos = util.removeDuplicateStringsInSmallArray(util.allCaseFirstLetters(prefix, 'full'));
+        // Put cap on exponential number of combinations (five words = 3^5 = 243 combinations)
+        search.type = fullCombos.length < 200 ? 'full' : 'basic';
+        // We have to remove duplicate string combinations because util.allCaseFirstLetters() can return some combinations
+        // where uppercase and lowercase combinations are exactly the same, e.g. where prefix begins with punctuation
+        // or currency signs, for languages without case, or where user-entered case duplicates calculated case
+        var prefixVariants = util.removeDuplicateStringsInSmallArray(
+            startArray.concat(
+                // Get basic combinations first for speed of returning results
+                util.allCaseFirstLetters(prefix).concat(
+                    search.type === 'full' ? fullCombos : []
+                )
+            )
+        );
         var dirEntries = [];
+
         function searchNextVariant() {
+            // If user has initiated a new search, cancel this one
+            if (globalstate.search.status === 'cancelled' || globalstate.search.prefix !== localPrefix)
+                return callback([], localPrefix);
             if (prefixVariants.length === 0 || dirEntries.length >= resultSize) {
-                callback(dirEntries);
-                return;
+                search.status = 'complete';
+                return callback(dirEntries, localPrefix);
             }
+            // Dynamically populate list of articles
+            search.status = 'interim';
+            if (!noInterim) callback(dirEntries, localPrefix);
             var prefix = prefixVariants[0];
             prefixVariants = prefixVariants.slice(1);
-            that.findDirEntriesWithPrefixCaseSensitive(prefix, resultSize - dirEntries.length, function (newDirEntries) {
-                dirEntries.push.apply(dirEntries, newDirEntries);
-                searchNextVariant();
-            });
+            that.findDirEntriesWithPrefixCaseSensitive(prefix, resultSize - dirEntries.length, localPrefix, 
+                function (newDirEntries, interim) {
+                    if (globalstate.search.status === 'cancelled' || globalstate.search.prefix !== localPrefix) 
```

Comment:
Using globalstate variable here breaks the software layers we have.
Using it in app.js is no problem (if it's where the variable is declared), but it should not be used in other files (except if passed as a parameter, of course).

The ZIM backend should not depend on external global variables. A drawback can be seen in the unit tests : you should be able to test each call independently, without being forced to declare the global variable (which might have side effects if the unit tests are ran in parallel, for example)